flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -0,0 +1,408 @@
+
+	private boolean initialized;
+
+	public StreamTwoInputSelectableProcessor(
+		Collection<InputGate> inputGates1,
+		Collection<InputGate> inputGates2,
+		TypeSerializer<IN1> inputSerializer1,
+		TypeSerializer<IN2> inputSerializer2,
+		Object lock,
+		IOManager ioManager,
+		StreamStatusMaintainer streamStatusMaintainer,
+		TwoInputStreamOperator<IN1, IN2, ?> streamOperator,
+		WatermarkGauge input1WatermarkGauge,
+		WatermarkGauge input2WatermarkGauge) {
+
+		checkState(streamOperator instanceof InputSelectable);
+
+		this.streamOperator = checkNotNull(streamOperator);
+		this.inputSelector = (InputSelectable) streamOperator;
+
+		this.lock = checkNotNull(lock);
+
+		InputGate unionedInputGate1 = InputGateUtil.createInputGate(inputGates1.toArray(new InputGate[0]));
+		InputGate unionedInputGate2 = InputGateUtil.createInputGate(inputGates2.toArray(new InputGate[0]));
+
+		// create a Input instance for each input
+		this.input1 = new StreamTaskNetworkInput(new BarrierDiscarder(unionedInputGate1), inputSerializer1, ioManager, 0);
+		this.input2 = new StreamTaskNetworkInput(new BarrierDiscarder(unionedInputGate2), inputSerializer2, ioManager, 1);
+
+		this.statusWatermarkValve1 = new StatusWatermarkValve(
+			unionedInputGate1.getNumberOfInputChannels(),
+			new ForwardingValveOutputHandler(streamOperator, lock, streamStatusMaintainer, input1WatermarkGauge, 0));
+		this.statusWatermarkValve2 = new StatusWatermarkValve(
+			unionedInputGate2.getNumberOfInputChannels(),
+			new ForwardingValveOutputHandler(streamOperator, lock, streamStatusMaintainer, input2WatermarkGauge, 1));
+
+		this.firstStatus = StreamStatus.ACTIVE;
+		this.secondStatus = StreamStatus.ACTIVE;
+
+		this.availableInputsMask = (int) new InputSelection.Builder().select(1).select(2).build().getInputMask();
+
+		this.lastReadInputIndex = 1; // always try to read from the first input
+
+		this.initialized = false;
+
+	}
+
+	public boolean processInput() throws Exception {
+		if (!initialized) {
+			initialize();
+		}
+
+		int readingInputIndex = inputSelection.fairSelectNextIndexOutOf2(availableInputsMask, lastReadInputIndex);
+		if (readingInputIndex == -1) {
+			return waitForAvailableInput(inputSelection);
+		}
+		lastReadInputIndex = readingInputIndex;
+
+		// if the input selection is ALL and availableInputsMask is not ALL,
+		// always try to check and set the availability of another input
+		if (availableInputsMask < 3 && inputSelection.isALLMaskOf2()) {
+			checkAndSetAvailable(1 - readingInputIndex);
+		}
+
+		StreamElement recordOrMark;
+		if (readingInputIndex == 0) {
+			recordOrMark = input1.pollNextNullable();
+			if (recordOrMark != null) {
+				processElement1(recordOrMark, input1.getLastChannel());
+			}
+		} else {
+			recordOrMark = input2.pollNextNullable();
+			if (recordOrMark != null) {
+				processElement2(recordOrMark, input2.getLastChannel());
+			}
+		}
+
+		if (recordOrMark == null && setUnavailableAndCheckFinished(readingInputIndex)) {
```

Comment:
I think this coupling of `setUnavailableInput` with `checkFinished()` is a bit difficult to follow and it would be better to have those two things separated:
1. have a one single call to `setUnavailableInput(readingInputIndex);`
2. followed by a isFinished check

```
if (recordOrMark == null) {
  setUnavailableInput(readingInputIndex);
}
return checkFinished();
```
Where 
```
boolean checkFinished() {
  boolean isFinished = input1.isFinished() && input2.isFinished();
  // set input selection accordingly
  return isFinished;
}
```
I think there is no performance penalty of always checking if either of input are finished, regardless of `readingInputIndex`, while it would simplify/generalize the code a little bit.